### PR TITLE
[SM-702] Extract access policy checks from create/update project commands

### PR DESCRIPTION
**Pull request by Thomas-Avery**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

The purpose of this PR is to extract request validation and authorization (access policy checks) from project create and update commands.

## Code changes

* bitwarden_license/src/Commercial.Core/SecretsManager/Commands/Projects/CreateProjectCommand.cs

Formatting 

* bitwarden_license/src/Commercial.Core/SecretsManager/Commands/Projects/UpdateProjectCommand.cs
src/Core/SecretsManager/Commands/Projects/Interfaces/IUpdateProjectCommand.cs

Extracting out access policy checks

* bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Projects/ProjectAuthorizationHandler.cs
bitwarden_license/src/Commercial.Core/SecretsManager/SecretsManagerCollectionExtensions.cs
src/Core/SecretsManager/AuthorizationRequirements/ProjectOperationRequirement.cs

New authorization handler and OperationAuthorizationRequirement

* bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/ProjectRepository.cs

Explicitly handle null case 

* bitwarden_license/test/Commercial.Core.Test/SecretsManager/Commands/AccessPolicies/CreateAccessPoliciesCommandTests.cs
bitwarden_license/test/Commercial.Core.Test/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommandTests.cs
bitwarden_license/test/Commercial.Core.Test/SecretsManager/Commands/AccessTokens/CreateAccessTokenCommandTests.cs
bitwarden_license/test/Commercial.Core.Test/SecretsManager/Commands/Projects/CreateProjectCommandTests.cs
bitwarden_license/test/Commercial.Core.Test/SecretsManager/Commands/Projects/DeleteProjectCommandTests.cs
bitwarden_license/test/Commercial.Core.Test/SecretsManager/Commands/Secrets/CreateSecretCommandTests.cs
bitwarden_license/test/Commercial.Core.Test/SecretsManager/Commands/Secrets/DeleteSecretCommandTests.cs
bitwarden_license/test/Commercial.Core.Test/SecretsManager/Commands/Secrets/UpdateSecretCommandTests.cs
bitwarden_license/test/Commercial.Core.Test/SecretsManager/Commands/ServiceAccounts/CreateServiceAccountCommandTests.cs
bitwarden_license/test/Commercial.Core.Test/SecretsManager/Commands/ServiceAccounts/RevokeAccessTokenCommandTests.cs
bitwarden_license/test/Commercial.Core.Test/SecretsManager/Commands/ServiceAccounts/UpdateServiceAccountCommandTests.cs
bitwarden_license/test/Commercial.Core.Test/SecretsManager/Commands/Trash/EmptyTrashCommandTests.cs
bitwarden_license/test/Commercial.Core.Test/SecretsManager/Commands/Trash/RestoreTrashCommandTests.cs

Reorganize test folder structure to match source code.

* bitwarden_license/test/Commercial.Core.Test/SecretsManager/Commands/Projects/UpdateProjectCommandTests.cs
bitwarden_license/test/Commercial.Core.Test/SecretsManager/Projects/UpdateProjectCommandTests.cs
test/Api.Test/SecretsManager/Controllers/ProjectsControllerTests.cs

Change unit tests to fit extracting access policy checks.

* bitwarden_license/test/Commercial.Core.Test/SecretsManager/AuthorizationHandlers/Projects/ProjectAuthorizationHandlerTests.cs

Unit tests for new authorization handler

* src/Api/SecretsManager/Controllers/ProjectsController.cs

Use authorization service to call new resource based authorization handlers.

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

**Review by MGibson1 — COMMENTED** [review]
This looks really good, and very in line with @eliykat's experimental https://github.com/bitwarden/server/pull/2845. I do think it's worth using the built-in asp.net `AuthorizationHandlers` to enable usage of other framework features like `[Authorize]` for a [Policy](https://learn.microsoft.com/en-us/aspnet/core/security/authorization/policies?view=aspnetcore-7.0) down the line.

**Review by MGibson1 — COMMENTED** [review]
Looks great, just two little questions

File: src/Core/SecretsManager/AuthorizationRequirements/ProjectOperationRequirement.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿using Microsoft.AspNetCore.Authorization.Infrastructure;
+
+namespace Bit.Core.SecretsManager.AuthorizationRequirements;
+
+public class ProjectOperationRequirement : OperationAuthorizationRequirement
+{
+}
+
+public static class ProjectOperations
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply src/Core/SecretsManager/AuthorizationRequirements/ProjectOperationRequirement.cs:9]: I was just following the pre-established convention in [the other PR](https://github.com/bitwarden/server/blob/33a7a6dc457e3f5c31d60650238c235b19b890e5/src/Core/OrganizationFeatures/AuthorizationHandlers/GroupOperations.cs). 

```
var authorizationResult = await _authorizationService.AuthorizeAsync(User, project, ProjectOperations.Update);
```

Vs  having it in the same class.

```
var authorizationResult = await _authorizationService.AuthorizeAsync(User, project, ProjectOperationRequirement.Update);
```

I don't have a preference, but do think we should stay consistent with it.

File: bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Projects/ProjectAuthorizationHandler.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+        switch (requirement)
+        {
+            case not null when requirement == ProjectOperations.Create:
+                await CanCreateProjectAsync(context, requirement, resource);
+                break;
+            case not null when requirement == ProjectOperations.Update:
+                await CanUpdateProjectAsync(context, requirement, resource);
+                break;
+        }
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Projects/ProjectAuthorizationHandler.cs:44]: Resolved in https://github.com/bitwarden/server/pull/2842/commits/2e548ed7cd51540a7b42653dd6a484c989381956

File: src/Core/SecretsManager/AuthorizationRequirements/ProjectOperationRequirement.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (MGibson1) [reply src/Core/SecretsManager/AuthorizationRequirements/ProjectOperationRequirement.cs:9]: I agree, just bringing it up because I don't know why it's that way in the other PR either 🤷

**Review by MGibson1 — APPROVED** [review]

File: src/Core/SecretsManager/AuthorizationRequirements/ProjectOperationRequirement.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (MGibson1) [reply src/Core/SecretsManager/AuthorizationRequirements/ProjectOperationRequirement.cs:9]: I'm happy doing that. It's a small change and not necessarily better.

**Comment by MGibson1** [comment]
@Thomas-Avery Can you please add a test to reflect each operation requirement and ensure none of them throw the unknown operation exception?

**Review by coltonhurst — DISMISSED** [review]
This looks great, assuming the test [mentioned](https://github.com/bitwarden/server/pull/2842#issuecomment-1520246229) covers the [cases](https://github.com/bitwarden/server/pull/2842#issuecomment-1520126224) @MGibson1 was raising. ✅

**Review by coltonhurst — APPROVED** [review]
The new changes added since the last review look good ✅

**Review by MGibson1 — APPROVED** [review]

**Review by MGibson1 — APPROVED** [review]
